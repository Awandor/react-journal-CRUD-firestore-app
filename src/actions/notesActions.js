// ========================================
// Notes Actions
// ========================================

import { db } from '../firebase/firebase-config';
import { loadNotes } from '../helpers/loadNotes';
import { types } from '../types/types';
import Swal from 'sweetalert2';
import { fileUpload } from '../helpers/fileUpload';

// Vamos a nombrar las funciones asíncronas con start...

export const startAddNewNoteAction = () => {

  // Como es una tarea asíncrona retornamos un callback, el primer argumento es dispatch aunque podemos ponerle el nombre que queramos
  // el segundo argumento es una función que retorna el state muy parecido a useSelect, también podemos usar el nombre que queramos
  // estos dos argumentos nos los da React

  return async( dispatch, getState ) => {

    // const state = getState();

    // console.log( state );

    const { uid } = getState().auth;

    // Creamos la nueva nota que vamos a guardar

    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    };

    try {

      // importamos la referencia a la base de datos de Firebase y usamos sus métodos

      // No hace falta definir la estructura previamente en Firebase, se genera automáticamente al guardar por primera vez

      const doc = await db.collection( `${uid}/journal/notes` ).add( newNote ); // add retorna una promesa

      dispatch( activeNoteStoreAction( doc.id, newNote ) );

      dispatch( addNewNoteStoreAction( doc.id, newNote ) );

    } catch ( err ) {

      console.log( 'error', err );

      Swal.fire( 'Error', 'FirebaseError', 'error' );

    }

  };

};

// No necesita exportarse pero sí para las pruebas

export const activeNoteStoreAction = ( id, note ) => ( {
  type: types.notesSetActive,
  payload: {
    id,
    ...note
  }
} );

export const addNewNoteStoreAction = ( id, note ) => ( {
  type: types.notesAddNewEntry,
  payload: {
    id,
    ...note
  }
} );

export const getNotesStoreAction = notes => ( {
  type: types.notesLoadEntries,
  payload: notes
} );

export const startLoadNotesAction = uid => {

  return async( dispatch ) => {

    // Cargamos las notas del usuario

    const notes = await loadNotes( uid ); // loadNotes retorna una promesa

    dispatch( getNotesStoreAction( notes ) );

  };

};

/* export const activateEntryAction = ( id ) => {

  return async( dispatch, getState ) => {

    const { notes } = getState().notes;

    console.log( 'activateEntryAction', notes );

    // Filtramos el arreglo para obtener el objeto que vamos a guardar como activo

    const activeNote = notes.filter( element => {

      return ( element.id === id ) || null;

    } );

    console.log( 'activeNote', activeNote[ 0 ] );

    // Creamos la nueva nota que vamos a setear en el Store como active



    // importamos la referencia a la base de datos de Firebase y usamos sus métodos

    // No hace falta definir la estructura previamente en Firebase, se genera automáticamente

    // const doc = await db.collection( `${uid}/journal/notes` ).add( newNote ); // add retorna una promesa

    // dispatch( activeNoteStoreAction( doc.id, newNote ) );

  };

}; */

export const startSaveNoteAction = ( note ) => {

  return async( dispatch, getState ) => {

    try {

      const { uid } = getState().auth;

      // Verificamos que url no sea undefined

      if ( !note.url ) {

        delete note.url;

      }

      // Lo clonamos para no modificar note, pues borramos el id

      const noteToFirestore = {...note };

      delete noteToFirestore.id; // Para Firestore no necesitamos el id

      await db.doc( `${uid}/journal/notes/${note.id}` ).update( noteToFirestore );

      // dispatch( startLoadNotesAction( uid ) ); // Esto carga de nuevo todo el panel lateral, sólo debería usarse para una paginación

      dispatch( refreshSidebarNoteStoreAction( note.id, note ) ); // necesitamos que la nota tenga el id, por eso no podemos usar noteToFirestore

      Swal.fire( 'Saved', note.title, 'success' );

    } catch ( err ) {

      console.log( 'error', err );

      Swal.fire( 'Error', 'FirebaseError', 'error' );

    }


  };

};

export const refreshSidebarNoteStoreAction = ( id, note ) => {

  return {
    type: types.notesUpdated,
    payload: {
      id,
      note
    }
  };

};

export const startImageUploading = ( file ) => {

  return async( dispatch, getState ) => {

    const { active: activeNote } = getState().notes;

    // console.log( file );

    // console.log( activeNote );



    try {

      Swal.fire( {
        title: 'Uploading',
        text: 'Please wait',
        showConfirmButton: false,
        allowOutsideClick: false,
        onBeforeOpen: () => {

          Swal.showLoading();

        }
      } );

      const fileUrl = await fileUpload( file );

      console.log( 'fileUrl', fileUrl );

      activeNote.url = fileUrl;

      dispatch( startSaveNoteAction( activeNote ) );

      Swal.fire( 'Saved', 'Image saved', 'success' );

    } catch ( err ) {

      console.log( 'error', err );

      Swal.fire( 'Error', 'FirebaseError', 'error' );

    }

  };

};

export const startDeleteNoteAction = ( id ) => {

  console.log( id );

  return async( dispatch, getState ) => {

    const uid = getState().auth.uid;

    // Delete from Firebase

    try {

      await db.doc( `${uid}/journal/notes/${id}` ).delete();

      dispatch( deleteNoteStoreAction( id ) );

      Swal.fire( 'Deleted', 'The note has been deleted', 'success' );

    } catch ( error ) {

      console.log( error );

      throw error;

    }


  };

};

export const deleteNoteStoreAction = ( id ) => ( {
  type: types.notesDeleteEntry,
  payload: { id }
} );

export const deleteAllNoteStoreAction = () => ( {
  type: types.notesLogoutCleaning
} );
