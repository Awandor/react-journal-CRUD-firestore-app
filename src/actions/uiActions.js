// ========================================
// UI Actions
// ========================================

import { types } from '../types/types';

// payload significa carga Ãºtil

export const setErrorAction = ( err ) => ( { type: types.uiSetError, payload: err } );

export const removeErrorAction = () => ( { type: types.uiRemoveError } );

export const startLoadingAction = () => ( { type: types.uiStartLoading } );

// export const finishLoadingAction = () => ( { type: types.uiFinishLoading } );

export const finishLoadingAction = () => {

  return { type: types.uiFinishLoading };

};
