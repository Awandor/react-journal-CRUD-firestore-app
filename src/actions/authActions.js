// ========================================
// Authentication Actions
// ========================================

import { types } from '../types/types';

import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { finishLoadingAction, startLoadingAction } from './uiActions';

import Swal from 'sweetalert2';
import { deleteAllNoteStoreAction } from './notesActions';

/* export const loginAction = ( uid, displayName ) => {

    return { type: types.login, payload: { uid, displayName } };

}; */

export const loginAction = ( uid, displayName ) => ( { type: types.login, payload: { uid, displayName } } );

export const startLoginEmailPasswordAction = ( email, password ) => {

  // La diferencia es que va a retornar un callback, esto es de la documentación de thunk, así se hace cualquier tarea asíncrona

  return ( dispatch ) => {

    // dispatch envía la acción a todos los reducers hasta que da con el que tiene el tipo adecuado que lo ejecuta cambiando el state

    // Aquí se pueden hacer varios dispatch

    /* setTimeout( () => {

      // Dispara una Acción síncrona una vez se resuelve la Acción asíncrona

      dispatch( loginAction( 123, 'Pedro' ) );

    }, 3500 ); */

    dispatch( startLoadingAction() );

    // Vamos a hacer que firebase retorne para las pruebas y ver los retornos de las acciones

    return firebase.auth().signInWithEmailAndPassword( email, password ).then( ( { user } ) => {

      dispatch( loginAction( user.uid, user.displayName ) );

      // Aquí podríamos realizar dispatch de varias cosas

      dispatch( finishLoadingAction() );

    } ).catch( err => {

      console.log( err );

      dispatch( finishLoadingAction() );

      Swal.fire( {
        title: 'Error!',
        text: err.message,
        icon: 'error',
        confirmButtonText: 'Close'
      } );

    } );

  };

};

export const startRegisterEmailPasswordNameAction = ( email, password, name ) => {

  // Como es una acción asíncrona retornamos un callback

  return ( dispatch ) => {

    return firebase.auth().createUserWithEmailAndPassword( email, password ).then( async( { user } ) => {

      // console.log( 'createUserWithEmailAndPassword', user );

      // Firebase no nos retorna el displayName, nos lo da como null

      // Con updateProfile podemos añadir el nombre, retorna una promesa que podemos manejar con then pero lo vamos a hacer con async await

      await user.updateProfile( { displayName: name } );

      dispatch( loginAction( user.uid, user.displayName ) );

      // Aquí podríamos realizar dispatch de varias cosas

    } ).catch( err => {

      console.log( err );

      Swal.fire( {
        title: 'Error!',
        text: err.message,
        icon: 'error',
        confirmButtonText: 'Close'
      } );

      throw err; // Para las pruebas estamos valorando el error

    } );

  };

};

export const startGoogleLoginAction = () => {

  // Como es una tarea asíncrona retornamos un callback que tiene dispatch como argumento que nos lo provee thunk

  return ( dispatch ) => {

    // Esto es una promesa

    firebase.auth().signInWithPopup( googleAuthProvider ).then( resp => {

      console.log( resp );

      dispatch( loginAction( resp.user.uid, resp.user.displayName ) );

    } );

  };

};

export const startLogoutAction = () => {

  // Como es una tarea asíncrona retornamos un callback que tiene dispatch como argumento que nos lo provee thunk

  return async( dispatch ) => {

    // Esto es una promesa

    await firebase.auth().signOut();

    dispatch( logout() );

    dispatch( deleteAllNoteStoreAction() );

  };

};

export const logout = () => ( { type: types.logout } ); // No es necesario export pero sí es necesario para las pruebas
