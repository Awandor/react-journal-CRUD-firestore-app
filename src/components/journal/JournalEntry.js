import React from 'react';

import moment from 'moment';
import { useDispatch } from 'react-redux';
import { /* activateEntryAction, */ activeNoteStoreAction } from '../../actions/notesActions';

export const JournalEntry = ( { id, date, title, body, url } ) => {

  const dateName = moment( date ).format( 'dddd' );

  const dateNumber = moment( date ).format( 'Do' );

  const dispatch = useDispatch();

  const handleEntryClick = () => {

    // Al tocar vamos a activar la nota pasando el id

    // dispatch( activateEntryAction( id ) );

    // No es necesario, podemos usar activeNoteStoreAction directamente

    dispatch( activeNoteStoreAction( id, { id, date, title, body, url } ) );

  };

  return (
    <div className="journal__entry" onClick={handleEntryClick}>

      {
        url && <div className="journal__entry-picture"
          style={
            { backgroundSize: 'cover', backgroundImage: `url(${url})` }}></div>
      }

      <div className="journal__entry-body">
        <p className="journal__entry-title">{title}</p>

        <p className="journal__entry-content">{body}</p>
      </div>

      <div className="journal__entry-date-box">
        <span>{dateName}</span><h4>{dateNumber}</h4>
      </div>
    </div>
  );

};
