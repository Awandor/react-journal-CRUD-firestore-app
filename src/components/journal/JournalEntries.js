import React from 'react';
import { useSelector } from 'react-redux';
import { JournalEntry } from './JournalEntry';

export const JournalEntries = () => {

    // const entries = [ 1,2,3,4,5,6,7,8,9,10 ];

    // Recuperamos las notas del Store

    const entries = useSelector( state => state.notes.notes );

    // console.log( entries );

    return (
        <div className="journal__entries">
            {
                entries.map( value => {

                    return <JournalEntry key={value.id} {...value} />;
                
                } )
            }
        </div>
    );

};
