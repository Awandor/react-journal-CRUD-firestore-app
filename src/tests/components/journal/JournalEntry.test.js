import { JournalEntry } from '../../../components/journal/JournalEntry';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';

// Como el componente hace uso del State necesitamos trabajar con el mock

import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from 'redux-thunk';
import { activeNoteStoreAction } from '../../../actions/notesActions';


const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

// Creamos el mock Store

const initialState = {};
  
let store = mockStore( initialState );

// Vamos a hacer un mock del dispatch

store.dispatch = jest.fn();


describe( 'Pruebas en <JournalEntry />', () => {

  const note = {
    id: '1234',
    date: 1609766312821,
    title: 'new title',
    body: 'new body',
    url: ''
  };
  const wrapper = mount( <Provider store={store}><JournalEntry key="5452535" {...note} /></Provider> );

  test( 'debería mostrarse correctamente', () => {

    expect( wrapper ).toMatchSnapshot();

  } );

  test( 'debería activar la nota', () => {
      
    wrapper.find( '.journal__entry' ).prop( 'onClick' )();

    expect( store.dispatch ).toHaveBeenCalled();

    expect( store.dispatch ).toHaveBeenCalledWith( activeNoteStoreAction( note.id, { ...note } ) );

  } );
  

} );
