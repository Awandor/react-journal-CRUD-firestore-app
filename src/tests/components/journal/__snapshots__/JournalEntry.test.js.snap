// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <JournalEntry /> deber√≠a mostrarse correctamente 1`] = `
<div
  className="journal__entry"
  onClick={[Function]}
>
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      new title
    </p>
    <p
      className="journal__entry-content"
    >
      new body
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Monday
    </span>
    <h4>
      4th
    </h4>
  </div>
</div>
`;
